### ISSUE TITLE
`Cstruct.hexdump` raises exception at Format module when called from multiple domains

### ISSUE THREAD

**Issue by Sudha247** [ocamllabs/ocaml-multicore#592]
Code:

```ocaml
let func = fun () ->
  for _ = 1 to 100 do
    let d = Cstruct.create 1000 in
    Cstruct.hexdump d
  done

let _ =
  let domains = Array.init 10 (fun _ -> Domain.spawn func) in
  func ();
  Array.iter Domain.join domains
```

dune
```
(executable
 (name cstruct_test)
 (modules cstruct_test)
 (libraries cstruct))
```

Backtrace:

```
$ OCAMLRUNPARAM=b _build/default/cstruct_test.exe
Fatal error: exception Stdlib.Queue.Empty
Raised at Stdlib__queue.take in file "queue.ml", line 73, characters 11-22
Called from Stdlib__format.advance_left in file "format.ml", line 436, characters 6-31
Called from Stdlib__format.kfprintf.(fun) in file "format.ml", line 1349, characters 16-34
Called from Stdlib__format.output_acc in file "format.ml", line 1298, characters 32-48
Called from Stdlib__format.output_acc in file "format.ml", line 1301, characters 32-48
Called from Stdlib__format.kfprintf.(fun) in file "format.ml", line 1349, characters 16-34
Called from Cstruct.hexdump_pp in file "lib/cstruct.ml", line 470, characters 4-62
Called from Stdlib__format.output_acc in file "format.ml", line 1288, characters 4-20
Called from Stdlib__format.kfprintf.(fun) in file "format.ml", line 1349, characters 16-34
Called from Dune__exe__Cstruct_test.func in file "cstruct_test.ml", line 4, characters 4-21
Called from Dune__exe__Cstruct_test in file "cstruct_test.ml", line 9, characters 2-9
```

**Comment by kayceesrk** [comment]
A clarification is that this bug appears even with domain-local format state introduced in https://github.com/ocaml-multicore/ocaml-multicore/pull/586.

I know what the problem is, but I don't know whether there is a nice solution. The crux of the problem is that both `printf` and `eprintf` get the domain-local formatters too early. `printf` and `eprintf` are implemented as follows in https://github.com/ocaml-multicore/ocaml-multicore/pull/586:

```ocaml
let printf fmt = fprintf (DLS.get std_formatter_key) fmt                             
let eprintf fmt = fprintf (DLS.get err_formatter_key) fmt
```

So, if you do:

```ocaml
let print_int_int = Format.printf "%d %d"
```
the type of the function `print_int_int` is `int -> int -> unit`. However, the domain-local formatter has already been chosen. Hence, if you run this function on a different domain, you will encounter the same error as the issue.

One possible fix is that you could do:

```ocaml
let print_int_int_fixed i j = Format.printf "%d %d" i j
```

which will have the same type as `print_int_int`, but delays the `DLS.get` until all the arguments are applied. Hence, this `print_int_int_fixed` will not raise the exception. 

The same issue happens in `CStruct.hexdump`: https://github.com/mirage/ocaml-cstruct/blob/master/lib/cstruct.ml#L474. If the code were changed to,

```ocaml
let hexdump v = Format.printf "@\n%a@." hexdump_pp v
```

the problem goes away. 

We should really fix stdlib to delay reading the domain-local formatter state until necessary to do so. CC @gasche.

**Comment by kayceesrk** [comment]
This should be fixed by https://github.com/ocaml-multicore/ocaml-multicore/pull/586/commits/a1e5baf371e2b8a13c7157735e5a810efe065e8f.

**Comment by kayceesrk** [comment]
Fixed by #586.
